### Информационно-поисковые системы
### Л/р №2. Выделение именованных сущностей

In [2]:
text = open('text.txt').read()
text[:500]

'Третий крестовый поход (1189—1192) был инициирован римскими папами Григорием VIII и (после смерти Григория VIII) Климентом III. В Крестовом походе приняли участие четверо самых могущественных европейских монархов — германский император Фридрих I Барбаросса из династии Штауфенов, французский король Филипп II Август — представитель династии Капетингов, австрийский герцог Леопольд V Бабенберг и английский король Ричард I Львиное Сердце из династии Плантагенетов. Также союзником европейских преслуг '

Итак. Статья из Вики про 3й крестовый поход. Примерный размер:

In [3]:
print(f'num of chars: {len(text)}\nnum of words: ~{len(text.split(" "))}')

num of chars: 54813
num of words: ~7637


Сущности будем вытаскивать следующие:
- даты: "в апреле 1191 года", "1189 год", "1096—1099"
- имена, в основном правителей Европы и Востока: "Генрих II Плантагенет", "Генрих Английский", "Салах-ад-дин", "король Ричард I Львиное Сердце из династии Плантагенетов"
- георгафия: "Малая Азия", "Акра"

Сначала посмотрим стандартные парсеры:

In [4]:
from natasha import NamesExtractor, DatesExtractor, LocationExtractor, PersonExtractor
from natasha.markup import show_markup, show_json

extractors = {'names': NamesExtractor(), 
              'persons': PersonExtractor(),
              'dates': DatesExtractor(), 
              'locations': LocationExtractor()
              }

matches = {k: v(text) for k, v in extractors.items()}

#spans = [_.span for _ in matches]
#facts = [_.fact.as_json for _ in matches]
#show_markup(text, spans)
#show_json(facts[:100])

Это даты и места:

In [5]:
import random 
facts = {k: [x.fact.as_json for x in m] for k, m in matches.items()}
for k in ['dates', 'locations']:
    print(f'--- {k} ---')
    show_json(random.choices(facts[k], k=30))

--- dates ---
[
  {
    "month": 8,
    "day": 23,
    "current_era": true
  },
  {
    "year": 1190,
    "current_era": true
  },
  {
    "year": 1191,
    "current_era": true
  },
  {
    "month": 8,
    "day": 23,
    "current_era": true
  },
  {
    "year": 1189,
    "current_era": true
  },
  {
    "year": 1191,
    "month": 1,
    "day": 20,
    "current_era": true
  },
  {
    "year": 1190,
    "month": 9,
    "current_era": true
  },
  {
    "year": 1190,
    "month": 3,
    "day": 25,
    "current_era": true
  },
  {
    "year": 1165,
    "current_era": true
  },
  {
    "year": 1188,
    "current_era": true
  },
  {
    "year": 1190,
    "month": 9,
    "current_era": true
  },
  {
    "year": 1190,
    "month": 10,
    "current_era": true
  },
  {
    "year": 1188,
    "current_era": true
  },
  {
    "year": 1189,
    "current_era": true
  },
  {
    "year": 1190,
    "month": 3,
    "day": 25,
    "current_era": true
  },
  {
    "month": 11,
    "day": 19,
    "current_er

In [6]:
facts = {k: [x.fact.as_json for x in m] for k, m in matches.items()}
persons = facts['persons']
for p in persons:
    p['name'] = ' '.join(f for f in p['name'].values())
persons = [' '.join(f.values()) for f in persons]
persons = list(set(persons))
print('--- persons ---')
print(sorted(persons))

--- persons ---
['Бароны и', 'Император не', 'Король ричард', 'После долгий', 'После смерти нур', 'Представителем жёсткого отношения к', 'Святой крест', 'август', 'айюбид', 'александр', 'алексей гид', 'алеппо', 'ангел', 'андроник кантакузин', 'ансберт', 'арсуф', 'асад', 'аскалон', 'аскалона франк', 'атлита', 'бабенберг', 'барбаросс', 'бертольд', 'бодуэн', 'больший', 'боэмунд', 'браничев', 'брат Ричарда иоанн', 'брата Саладина малёк', 'вильгельм', 'вильгельм завоеватель', 'вильгельм тирский', 'гвидо', 'гвискара', 'генрих', 'генрих английский', 'георг роппельт', 'главе отличная', 'графом Анжу мэн и', 'графы Анжу мэн и', 'губернатору Браничева и', 'дин', 'дочерью бертольд герцог', 'дочь Рожера и', 'дочь английского генрих король', 'дюрнштейн', 'жан', 'император не', 'император фридрих', 'император фридрих германский', 'императора Исаака ангел', 'императора Исаака избраннейших и', 'императора в живой', 'императора то', 'императором и славянский', 'иоанн', 'иоанн дук', 'исаак', 'исаак комни

In [7]:
names = facts['names']
names = [' '.join(f.values()) for f in names]
baseline_names = list(set(names))
print(f'--- names ---')
print(sorted(baseline_names))

--- names ---
['август', 'айюбид', 'александр', 'алексей гид', 'алеппо', 'алиса', 'ангел', 'андроник кантакузин', 'ансберт', 'арсуф', 'асад', 'аскалон', 'аскалона франк', 'атлита', 'бабенберг', 'барбаросс', 'беренгарий', 'бертольд', 'бодуэн', 'больший', 'боэмунд', 'браничев', 'вильгельм', 'вильгельм завоеватель', 'вильгельм тирский', 'гвидо', 'гвискара', 'генрих', 'генрих английский', 'георг роппельт', 'григорий', 'дин', 'дюрнштейн', 'жан', 'иоанн', 'иоанн безземельный', 'иоанн дук', 'ираклий', 'исаак', 'исаак комнин', 'климент', 'конрад', 'конрад монтферратский', 'конрад монферратский', 'лампрон', 'левон', 'леопольд', 'лузиньян', 'львиное', 'малая', 'малёк', 'месопотамия', 'наваррская', 'неманя', 'никита', 'нур', 'петра', 'плантагевший', 'птолемаида', 'пётр', 'раймунд', 'райнальда', 'ричард', 'роберт', 'роберт сент', 'романия', 'саладин', 'саладинова', 'себастьен мамро', 'селевкий', 'сирия', 'стефан', 'тарс', 'филипп', 'фрагонар', 'фридрих', 'фридрих германский', 'фридрих швабский', '

В общем... 
- с датами проблем вроде нет, парсится хорошо, даже цепляет тройки день-месяц-год
- локации тоже неплохо вытаскивает, благо они все однословные
- а вот с именами все сильно хуже. Persons-парсер вообще выдает шума больше, чем полезного, а именной парсер явно не приспособлен извлекать заковыристые средневеконизмы, да и на титулы он явно не рассчитан.

Вот доп. парсингом имен дальше и займемся

In [8]:
from yargy import (
    Parser,
    rule, or_, and_, not_
)
from yargy.predicates import (
    eq, in_,
    type, normalized,
    dictionary,
    gte, lte, is_capitalized
)
from yargy.pipelines import (
    pipeline,
    morph_pipeline
)
from yargy.interpretation import (
    fact,
    attribute
)
from yargy.predicates import gram
from yargy.relations import gnc_relation

gnc = gnc_relation() 

**Правило первое** - FunnyName. Будем парсить имена по французскому и восточному стилю в духе "Салах-ад-Дин" или "Рено де Шатильон". Соответственно, кастом на виды разделителей (- или space) и "среднее имя".

По результату: нашел много салахов и нур'а... нашел бы и больше, если бы стандартная распознавашка имен признавала за имена/фамилии побольше всего нетипичного

In [48]:
FunnyName = fact(
    'Name',
    ['first', 'mid', 'last']
)


NAME_SEP = eq('-')
MID_NAME = or_(
    eq('де').interpretation(FunnyName.mid.const('де')), 
    eq('ад').interpretation(FunnyName.mid.const('ад'))
)
NAME_FIRST = gram('Name').interpretation(
    FunnyName.first.inflected()
).match(gnc)
NAME_LAST = or_(gram('Name'), gram('Surn')).interpretation(
    FunnyName.last.inflected()
).match(gnc)


FUNNY_NAME = or_(
    rule(
    NAME_FIRST, NAME_SEP,
    MID_NAME, NAME_SEP,
    NAME_LAST), 
    rule(
    NAME_FIRST,
    MID_NAME,
    NAME_LAST)
).interpretation(
    FunnyName
)

parser1 = Parser(FUNNY_NAME)
print('--- rule #1, test:')
test_text = 'Салах-ад-Дин, Салах ад дин, Иван де Павлов'
for match in parser1.findall(test_text):
    start, stop = match.span
    print(match.fact, test_text[start:stop])

print('\n\n--- rule #1, real text:')
funny_names = list(parser1.findall(text))
for match in list(funny_names):
    start, stop = match.span
    print(text[start:stop])

--- rule #1, test:
Name(first='салах', mid='ад', last='дин') Салах-ад-Дин
Name(first='салах', mid='ад', last='дин') Салах ад дин
Name(first='иван', mid='де', last='павлов') Иван де Павлов


--- rule #1, real text:
Салах-ад-дин
Салах-ад-дин
Салах-ад-дин
Салах-ад-дин
Нур-ад-дин
Салах-ад-дин
Салах-ад-дин
Салах-ад-дин
Салах-ад-дин
Салах-ад-дин
Салах-ад-дин


**Правило второе** - RoyalName. Поиск более сложных имен, которые начинаются таки с имени, а дальше могут включать в себя римский номер ("Генрих VI"), а также какие-то стремные слова (0 или больше), которые начинаются с большой буквы, но стандартным парсером за имена и фамилии не считаются... да могут ими и не быть ("Филипп IV Красивый")

In [49]:
import roman

RoyalName = fact(
    'Name',
    ['first_name', 'regnal_number', 'last_name']
)

RNAME_FIRST = and_(is_capitalized(), gram('Name')).interpretation(
    RoyalName.first_name.inflected()
).match(gnc)
REGNAL_NUMBER = pipeline([roman.toRoman(x) for x in range(30)]).interpretation(
    RoyalName.regnal_number
)
RNAME_LAST = or_(
    gram('Name'), gram('Surn'),
    #and_(is_capitalized(), (gram('NOUN'))),
).interpretation(
    RoyalName.last_name.inflected()
).match(gnc)

RNAME_LAST_EXTENDED = or_(
    gram('Name'), gram('Surn'),
    is_capitalized(),
).interpretation(
    RoyalName.last_name.inflected()
).match(gnc)

ROYAL_NAME = or_(rule(RNAME_FIRST, REGNAL_NUMBER, RNAME_LAST_EXTENDED.optional().repeatable(max=3)),
                 rule(RNAME_FIRST, REGNAL_NUMBER.optional(), RNAME_LAST)
).interpretation(RoyalName)

parser2 = Parser(ROYAL_NAME)
print('--- rule #2 test:')
test_text = 'Александр XIII, Пиппин IV Мохнатые Ноги, Петр, Конрад Зеленый, Конрад V Зеленый'
for match in parser2.findall(test_text):
    start, stop = match.span
    print(match.fact, test_text[start:stop])

print('\n\n--- rule #2, real text:')
royal_names = list(parser2.findall(text))
for match in royal_names:
    start, stop = match.span
    print(text[start:stop])


--- rule #2 test:
Name(first_name='александр', regnal_number='XIII', last_name=None) Александр XIII
Name(first_name='пиппина', regnal_number='IV', last_name='нога') Пиппин IV Мохнатые Ноги
Name(first_name='конрад', regnal_number='V', last_name='зелёный') Конрад V Зеленый


--- rule #2, real text:
Григорием VIII
Григория VIII
Климентом III
Фридрих I
Филипп II Август
Леопольд V Бабенберг
Ричард I
Левон II
Асад ад
Асад ад
Асад ад
Салах ад
Вильгельма Тирского
Роберт Сент
Себастьена Мамро
Фридрих I
Филипп II Август
Генрих II Плантагенет
Филипп II Август
Генрих II
Генриху II
Генрих II
Вильгельма II
Генриха VI
Вильгельмом II
Генриха II
Исаака Комнина
Исаак Комнин
Исаака Комнина
Фридрих I
Исаак Ангел
Фридрих I
Бела III
Фридриху I
Роберт Гвискар
Никиты Хониата
Андроника Кантакузина
Исаака Ангела
Исаака Ангела
Бела III
Фридриха I
Фридриха I
Исаака Ангела
Себастьена Мамро
Конрадом Монтферратским
Эвариста Фрагонара
Леопольд V
Филипп Август
Конрада Монтферратского
Филиппу II Августу
Леопольд V
Леоп

**Правило третье** - TitledName. Ищем имена с титулами. Имена берутся как комбо предыдущих правил, к ним добавляется поиск заданного титула.

In [36]:
TitledName = fact(
    'TitledName',
    ['title', 'name']
)

TITLE = morph_pipeline([
    'король', 'герцог' 'халиф', 'лорд', 'барон', 'император',
    'граф', 'визирь'
]).interpretation(
    TitledName.title.inflected()
).match(gnc)

JUST_NAME = gram('Name').interpretation(
    TitledName.name.inflected()
).match(gnc)

TITLED_NAME = rule(
    TITLE, 
    or_(FUNNY_NAME, ROYAL_NAME, JUST_NAME).interpretation(TitledName.name.inflected()).match(gnc)
).interpretation(TitledName)


parser3 = Parser(TITLED_NAME)
print('--- rule #3 test:')
test_text = 'король Конрад V Зеленый, граф Анджей, визирь Иван-ад-дин'
for match in parser3.findall(test_text):
    start, stop = match.span
    print(test_text[start:stop])

print('\n\n--- rule #3, real text:')
titled_names = list(parser3.findall(text))
for match in titled_names:
    start, stop = match.span
    print(text[start:stop])


--- rule #3 test:
король Конрад V Зеленый
граф Анджей
визирь Иван-ад-дин


--- rule #3, real text:
император Фридрих I
король Филипп II Август
король Ричард I
король Филипп II Август
король Генрих II
короля Генриха II
Король Ричард
император Фридрих I
императора Фридриха
императора Исаака
императора Исаака Ангела
королю Ричарду
король Иоанн


### Итого
- Стандартные правила позволили неплохо выделить даты и локации, но у них плохо вышло выделить составные имена, а парсер Persons (ради титулов и прочего) выдавал слишком много шума
- Собственные правила направлены на улучшенный поиск имен, а именно:
  1. Именные приставки и что-то похожее: Салах-ад-Дин, Жак де Моле
  2. Королевские имена с номерами и прозвищами: Пиппин III Длинная Трубка
  3. Титулы перед именами: король <...> 

Сформируем итоговый набор сущностей

In [16]:
dates = [f'date: ' + ', '.join(f"{k}={v}" for k, v in fact.items()) for fact in facts['dates']]
locations = [f'location: ' + ', '.join(f"{k}={v}" for k, v in fact.items()) for fact in facts['locations']]

print(dates[0], '\n', locations[0], sep='')

date: year=1187, month=10, current_era=True
location: name=киликийская армения


In [51]:
funny_names_str = [f'name [rule1]:' + ', '.join(f"{k}={rn.fact.__dict__[k]}" for k in rn.fact.__dict__ if k != '_raw') for rn in funny_names]
royal_names_str = [f'name [rule2]:' + ', '.join(f"{k}={rn.fact.__dict__[k]}" for k in rn.fact.__dict__ if k != '_raw') for rn in royal_names]


print(funny_names_str[0], '\n', royal_names_str[0], sep='')


name [rule1]:first=салах, mid=ад, last=дин
name [rule2]:first_name=григорий, regnal_number=VIII, last_name=None


In [59]:
titled_names_str = [f'name [rule3]: ' + f'title={tn.tokens[0].value}, name=' + ' '.join([t.value for t in tn.tokens[1:]])  for tn in titled_names]

titled_names_str[10]

'name [rule3]:title=императора, name=Исаака Ангела'

In [62]:
with open('extracted_names.txt', 'w') as f:
    str = '\n========\n'.join('\n'.join(data) for data in [dates, locations, funny_names_str, royal_names_str, titled_names_str])
    f.write(str)